python/GafferUI/PlugCreationWidget.py

```diff
@@ -0,0 +1,361 @@
+	# to create plugs. The signature for slots is `( menuDefinition, plugCreationWidget )`,
+	# and slots should just modify the menu definition in place.
+	@staticmethod
+	def plugCreationMenuSignal() :
+
+		return PlugCreationWidget.__plugCreationMenuSignal
+
+	## Signal emitted to allow customisation of the drag & drop behaviour.
+	# The signature for slots is `( plugCreationWidget, dragDropEvent )` and
+	# they should return a callable which will be called with the same
+	# arguments on drop.
+	@staticmethod
+	def plugCreationDragEnterSignal() :
+
+		try :
+			return PlugCreationWidget.__plugCreationDragEnterSignal
+		except AttributeError :
+			PlugCreationWidget.__plugCreationDragEnterSignal = Gaffer.Signals.Signal2(
+				PlugCreationWidget.__dragEnterSignalCombiner
+			)
+			return PlugCreationWidget.__plugCreationDragEnterSignal
+
+	## Creates a plug on `plugParent()`. Expected to be called from menu items
+	# created by `plugCreationMenuSignal()`. The optional `name` argument is
+	# used only when wrapping the created plug in a NameValuePlug or TweakPlug.
+	def createPlug( self, prototypePlug, name = "" ) :
+
+		plug = prototypePlug.createCounterpart( prototypePlug.getName(), Gaffer.Plug.Direction.In )
+
+		action = Gaffer.Metadata.value( self.__plugParent, "plugCreationWidget:action" ) or "addPlug"
+
+		if action == "addPlug" and isinstance( self.__plugParent, Gaffer.CompoundDataPlug ) :
+			action = "addNameValuePlug"
+		elif action == "addPlug" and isinstance( self.__plugParent, Gaffer.TweaksPlug ) :
+			action = "addTweakPlug"
+
+		with Gaffer.UndoScope( self.__plugParent.ancestor( Gaffer.ScriptNode ) ) :
+
+			match action :
+				case "addPlug" :
+					if isinstance( self.__plugParent, Gaffer.Box ) :
+						## \todo Could this be made the default via a metadata registration in SubGraphUI.py?
+						Gaffer.Metadata.registerValue( plug, "nodule:type", "" )
+					plug.setFlags( Gaffer.Plug.Flags.Dynamic, True )
+					self.__plugParent.addChild( plug )
+				case "addNameValuePlug" :
+					plug = Gaffer.NameValuePlug( "", plug, True, "member0", flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic )
+					plug["name"].setValue( name )
+					self.__plugParent.addChild( plug )
+				case "addTweakPlug" :
+					plug = Gaffer.TweakPlug( "tweak0", valuePlug = plug, flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic )
+					plug["name"].setValue( name )
+					self.__plugParent.addChild( plug )
+				case "setup" :
+					self.__plugParent.setup( plug )
+				case "addQuery" :
+					self.__plugParent.node().addQuery( plug, name )
+
+	def __menuDefinition( self ) :
+
+		result = IECore.MenuDefinition()
+
+		pendingDivider = None
+		def appendItem( menuPath, plugType, plugKW = {} ) :
+
+			if not self.__plugTypeIncluded( plugType ) :
+				return
+
+			nonlocal pendingDivider
+			if pendingDivider and result.size() :
+				result.append( pendingDivider, { "divider" : True } )
+				pendingDivider = None
```

Comment:
I feel the logic here is not quite right. If you were to register `"plugCreationWidget:includedTypes", "Gaffer.Color3fPlug Gaffer.Color4fPlug"` you'd end up with the ColorDivider between the Color3f and Color4f menu items.

Maybe we should be resetting `pendingDivider` to `None` when creating the first actual menu item?